### DISCUSSION TITLE
Tasmota with DDP and WLED

### DISCUSSION THREAD

**Discussion by kirintala** [arendst/Tasmota#20289] [Q&A]
Does the current Tasmota build include support for DDP-protocol? Ive found a separate build from Cloudfree website with DDP support but cannot get WLED controlling my Tasmota RGB controller (H801) via DDP.
Do you know if this is possible? And of course how :) ?

**Comment by barbudor** [comment]
I have no idea how it works but you can read those PR:
https://github.com/arendst/Tasmota/pull/13937
https://github.com/arendst/Tasmota/pull/14017

The support for WS2812DDP requires `#define USE_NETWORK_LIGHT_SCHEMES`
This seems to be included by default in ESP32 variants but not with ESP8266
To use in a H801 you need to compile your own binary with the above added in `user_config_override.h`

**Reply by barbudor** [reply]
Note the article on Cloudfree is almost 2 years old so the binary provided there is likely as old
While it is unlikely anything as changed on DDP support, it may still be better to get yourself a fresher one.

If you don't succeed, you may try to ping the author @ tony-fav


For reference, the cloudfree blog article: https://cloudfree.shop/blog/2022/02/17/pair-tasmota-with-wled-for-lighting-effects/

**Reply by sfromis** [reply]
If you tell details of how you tried making it work, *maybe* someone can spot some "opportunity for improvement". This page may be useful for more general info: http://www.3waylabs.com/ddp/

**Reply by kirintala** [reply]
Just plain and simple:
- installed the tasmota-DDP.bin from Cloudfree site to my H801 controller
- Have WLED installed to another controller controlling some addressable LEDS (works nicely)
- Configured the WLED to include the IP of the H801 controller as DDP RGBW additional to the addressable two LED strips.

Thats more or less it.  When WLED has one of it's LED-strips as DDP RGBW, with the correct IP-address, should it then control the ON/OF and color-states of that DDP strip?

**Reply by sfromis** [reply]
You did not configure the H801 then? Do not assume that the Cloudfree binary is configured for the specific hardware you misused it on.

**Reply by kirintala** [reply]
I did configure it by issuing Scheme 5 -command in the command line. The Tasmota & H801 -controller itself work fine, ON/OFF, colors, dimming etc. And actually the Scheme 5 -command shows out that tasmota starts UPD-listening for PPD.  Ive also checked that my UDP-ports/comminuvcation is not restricted in my router.
It just does not work and receive commands from WLED.

I am not assuming anything and not misusing anything. I can stop asking for help here if it bother you ?

**Reply by Jason2866** [reply]
the h801 is configured like a RGB bulb?

**Reply by sfromis** [reply]
I was not "bothered", just bemused by an apparent expectation that using a binary customized for a completely different device as "the way" to get DDP working. The word "misuse" was not to bother you, what I wrote now was the only meaning.

What is important is details of how the device was configured, and you told nothing about that, nor of what you did to try to make it work on the H801. And I have no info about how the build you used even is configured, or maybe even modified.

In lots of cases, getting exact details of what was done allows to spot something missing. The thing is that most questions boils down to the omission of some significant detail.

Anyway, doing a quick test with a WLED device configured for DDP output, and having a Tasmota device with a custom build including DDP, `Scheme 5`, I did see some "life" in the neopixels after powering off/on in WLED. I've not used WLED/DDP enough to be sure of it working as it "should", but I did get flashing colors. It was not with a bulb, but with a neopixel strip; still I saw all pixels same color, bulb-like.

Reconfiguring the Tasmota board with 5 PWM channels, bulb-like similar to H801, instead of the neopixel strip, I did get similar activity on those channels. Again "signs of life", but only after explicitly having the light in Tasmota turned on.

**Comment by Jason2866** [comment]
For Neopixels Scheme 5 is wrong.

**Reply by sfromis** [reply]
In my test, `Scheme 5` allowed control from WLED of the Tasmota light. Looks like the docs is wrong, or the code was changed in an incompatible way to have `Scheme 5` for DDP instead.

**Reply by arendst** [reply]
If define USE_NETWORK_LIGHT_SCHEMES is enabled all schemes above 4 move up by one ;-)

**Reply by sfromis** [reply]
Users may not know if `USE_NETWORK_LIGHT_SCHEMES` is in force, hence it gets confusing when the numbers are not working as expected. Specifically, @Jason2866 has done the nice thing of enabling the feature in "special" builds, meaning that when using `tasmota32s2-lvgl.bin` (and so on), I'm not seeing the `Scheme` command work as in the docs, but instead the undocumented lift of numbers 5-14. Way to confound people :wink:

**Reply by Jason2866** [reply]
@sfromis It is enabled in all esp32x builds https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L1122

**Reply by arendst** [reply]
The fact it's scheme 5 is caused by the possiblity of NOT having the ws2812 driver enabled. The DDP code owner had to find a way to add a scheme in an consequtive order even when the WS2812 driver isn't there.

I agree it's an unlucky choice but as time goes by new features are added and some of them just can't be added nicely ;-)

**Reply by sfromis** [reply]
Added to docs:
> **Note:** If using ESP32, or any build with the DDP feature (`USE_NETWORK_LIGHT_SCHEMES`), enabling DDP is `Scheme 5`, and the normal neopixel scheme numbers become 6-15 instead of 5-14

Now we can at least point confounded users to something.....

**Reply by jkwim** [reply]
I compiled in `USE_NETWORK_LIGHT_SCHEMES` in a build for ESP8266.

Integrated with WLED.

Configured a 10-pixel WS28212 on Tasmota

`Scheme 5 `activated.

DDP works but all 10 pixels in WS2812 strip get lit up with the same color. No animations continue in to the WS2812 strip.

So there is an issue of translating DDP data stream in to WS2812 data stream.

I have a small video clip.


https://github.com/arendst/Tasmota/assets/84637540/77ba65f7-f2ab-476d-bd5d-d62d35800b3d

**Reply by sfromis** [reply]
Getting same color on all pixels of the strip is consistent what I mentioned 4 days ago, also when the neopixel count in WLED is lower than in Tasmota. Action looks like it was treated as a "bulb" instead of being processed as addressable leds.

**Reply by sfromis** [reply]
With ESP32, I've found that it works better if you use `Scheme 16` instead of `Scheme 5`

**Reply by jkwim** [reply]
So probably the implementation is similar to that of OpenBK. I have tried 3 x RGB bulbs (3 x different IPs) running OpenBK with WLED and they work fine. Each bulb is a single device.

**Reply by sfromis** [reply]
Yeah, sounds like it is similar when it is `Scheme 5`, with `Scheme 16` instead using the full features with addressable leds.
